sfdc api username and field security clarifications (#1184)
rudderlabs/rudderstack-docs: docs/destinations/streaming-destinations/salesforce.mdx

<img src="../../assets/event-stream-destinations/salesforce-connection-settings.png" alt="Salesforce connection settings" />

- **User Name**: Enter your Salesforce username in this field.

> <div class="warningBlock">The salesforce API does not support certain special characters in the username. For example, if you create a secondary user with the email <code class="inline-code">john.smith+rudder@domain.com</code>, Salesforce will set the username to that email address by default. In order to work with the Salesforce API you will need to remove the <code class="inline-code">+</code> sign from the username.</div>

Reviewer: ```suggestion
<div class="warningBlock">
**The salesforce API does not support certain special characters in the username.**

For example, if you create a secondary user with the email <code class="inline-code">john.smith+rudder@domain.com</code>, Salesforce sets the username to that email address by default. However, to work with the Salesforce API you will need to remove the <code class="inline-code">+</code> sign from the username.
</div>
```

RudderStack looks for the key `externalId` under `context` and determines the Salesforce object type by removing the part `Salesforce-` from the field `type`. Further, it makes a `PATCH` request if there is an `id` present in the request to update the record. Otherwise, a new record is created.

> <div class="warningBlock">You cannot send an <code class="inline-code">id</code> property when creating new records. This will cause an API failure. When creating new records only provide the <code class="inline-code">type: "Salesforce-Contact"</code> portion of the <code class="inline-code">externalId</code> object.</div>

Reviewer: ```suggestion
<div class="warningBlock">
You cannot send an <code class="inline-code">id</code> property when creating new records as it  will cause an API failure.

When creating new records, only provide the <code class="inline-code">type: "Salesforce-Contact"</code> portion of the <code class="inline-code">externalId</code> object.
</div>
```

### How do I check the number of Salesforce API calls left for the day?

To check the number of Salesforce API calls, go to **Setup** > **Administration Setup** > **Company Profile** > **Company Information**. You should then be able to see a field called **API Requests, Last 24 Hours**, which contains the number of API calls left for the day.

### How to fix "No such column 'X' on sobject of type Y" errors

> If event delivery to Salesforce fails due to a provided field not existing, even though the field does exist, check the Field Level Security settings of that field in Salesforce. When the field is **not** marked as visible to the role your RudderStack Salesforce user is using, this error is thrown. To fix it, just make the field visible to the appropriate role.

Reviewer: ```suggestion
If event delivery to Salesforce fails due to a provided field not existing even though it exists in the event, check the field-level security settings for that particular field in Salesforce.

You will get this error when the field is **not** marked as visible to the role your RudderStack Salesforce user is using. To fix it, make the field visible to the appropriate role.
```

Reviewer: thanks for the clarification - @maximum-pixels - this is crystal clear now.

Reviewer: Updated the above suggestion accordingly - please let me know if it looks good.
